backend/gce.go

```
@@ -868,42 +912,90 @@ func (i *gceInstance) RunScript(ctx gocontext.Context, output io.Writer) (*RunRe
+	select {
+	case err := <-c.errChan:
+		return err
+	case <-ctx.Done():
+		if ctx.Err() == gocontext.DeadlineExceeded {
+			metrics.Mark("worker.vm.provider.gce.delete.timeout")
+		}
+		return ctx.Err()
+	}
+}
+
+func (i *gceInstance) stepDeleteInstance(c *gceInstanceStopContext) multistep.StepAction {
 	op, err := i.client.Instances.Delete(i.projectID, i.ic.Zone.Name, i.instance.Name).Do()
 	if err != nil {
-		return err
+		c.errChan <- err
+		return multistep.ActionHalt
 	}
 
-	errChan := make(chan error)
-	go func() {
-		zoneOpCall := i.client.ZoneOperations.Get(i.projectID, i.ic.Zone.Name, op.Name)
+	c.instanceDeleteOp = op
+	return multistep.ActionContinue
+}
 
-		for {
-			newOp, err := zoneOpCall.Do()
-			if err != nil {
-				errChan <- err
-				return
-			}
+func (i *gceInstance) stepWaitForInstanceDeleted(c *gceInstanceStopContext) multistep.StepAction {
+	logger := context.LoggerFromContext(c.ctx)
 
-			if newOp.Status == "DONE" {
-				if newOp.Error != nil {
-					errChan <- &gceOpError{Err: newOp.Error}
-					return
-				}
+	logger.WithFields(logrus.Fields{
+		"duration": i.ic.StopPrePollSleep,
+	}).Debug("sleeping before first checking instance delete operation")
 
-				errChan <- nil
-				return
-			}
+	time.Sleep(i.ic.StopPrePollSleep)
+
+	zoneOpCall := i.client.ZoneOperations.Get(i.projectID,
+		i.ic.Zone.Name, c.instanceDeleteOp.Name)
+
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = i.ic.StopPollSleep
+	b.MaxInterval = 10 * i.ic.StopPollSleep
+	b.MaxElapsedTime = 2 * time.Minute
 
-			time.Sleep(i.provider.bootPollSleep)
+	err := backoff.Retry(func() (err error) {
+		newOp, err := zoneOpCall.Do()
+		if err != nil {
+			return
 		}
-	}()
 
-	select {
-	case err := <-errChan:
-		return err
-	case <-ctx.Done():
-		return ctx.Err()
+		if newOp.Status == "DONE" && newOp.Error == nil {
+			err = nil
+			return
+		}
+
+		if newOp.Error != nil {
+			err = &gceOpError{Err: newOp.Error}
+			return
+		}
+
+		return
```

We should probably return an error in this case, since we'd like it to retry if we get this far?